lib/ansible/plugins/action/gather_facts.py

```diff
@@ -0,0 +1,110 @@
+        parallel = task_vars.pop('ansible_facts_parallel', self._task.args.pop('parallel', None))
+
+        if 'smart' in modules:
+            connection_map = C.config.get_config_value('CONNECTION_FACTS_MODULES', variables=task_vars)
+            modules.extend([connection_map.get(self._connection._load_name, 'setup')])
+            modules.pop(modules.index('smart'))
+
+        failed = {}
+        skipped = {}
+        if parallel is False or (len(modules) == 1 and parallel is None):
+            # serially execute each module
+            for fact_module in modules:
+                # just one module, no need for fancy async
+                mod_args = self._get_module_args(fact_module, task_vars)
+                res = self._execute_module(module_name=fact_module, module_args=mod_args, task_vars=task_vars, wrap_async=False)
+                if res.get('failed', False):
+                    failed[fact_module] = res.get('msg')
+                elif res.get('skipped', False):
+                    skipped[fact_module] = res.get('msg')
+                else:
+                    result = combine_vars(result, {'ansible_facts': res.get('ansible_facts', {})})
+        else:
+            # do it async
+            jobs = {}
+            for fact_module in modules:
+
+                mod_args = self._get_module_args(fact_module, task_vars)
+                self._display.vvvv("Running %s" % fact_module)
+                jobs[fact_module] = (self._execute_module(module_name=fact_module, module_args=mod_args, task_vars=task_vars, wrap_async=True))
+
+            while jobs:
+                for module in jobs:
+                    poll_args = {'jid': jobs[module]['ansible_job_id'], '_async_dir': os.path.dirname(jobs[module]['results_file'])}
+                    res = self._execute_module(module_name='async_status', module_args=poll_args, task_vars=task_vars, wrap_async=False)
+                    if 'finished' in res:
+                        if res.get('failed', False):
+                            failed[module] = res.get('msg')
+                        elif res.get('skipped', False):
+                            skipped[module] = res.get('msg')
+                        else:
+                            result = combine_vars(result, {'ansible_facts': res.get('ansible_facts', {})})
+                        del jobs[module]
+                        break
+                else:
+                    time.sleep(0.5)
+
+        if skipped:
+            result['msg'] = "The following modules were skipped: %s\n" % (', '.join(skipped.keys()))
+            for skip in skipped:
+                result['msg'] += '  %s: %s\n' % (skip, skipped[skip])
+            if len(skipped) == len(modules):
+                result['skipped'] = True
+
+        if failed:
+            result['failed'] = True
+            result['msg'] += "The following modules failed to execute: %s\n" % (', '.join(failed.keys()))
```

Comment:
This would result in KeyError. It needs to be `result['msg'] =` instead of `result['msg'] +=`.